File: LeetCode/Cpp/1-two-sum.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
    	vector<int> result;
        for (int i = 0; i < nums.size(); i ++) {
            if ((nums[i] > target && nums[i] > 0) || (nums[i] < target && nums[i] < 0)) {
                continue;
            }

            for (int j = i+1; j < nums.size(); j++) {
                if ((nums[i] + nums[j]) == target) {
                    result.push_back(i);
                    result.push_back(j);
                }
            }
        }
        return result;
    }
};

int main() {
    Solution* solution = new Solution;
    int inputArray[4] = {2, 7, 11, 15};
    vector<int> inputVector;

    for (int i=0; i<sizeof(inputArray)/sizeof(inputArray[0]); i++) {
    	inputVector.push_back(inputArray[i]);
	}

	for (int i=0; i<solution->twoSum(inputVector, 9).size(); i++) {
		cout<<solution->twoSum(inputVector, 9)[i];
	}
}

File: LeetCode/Java/1-two-sum.java
/**
 * Description
 * Given an array of integers, return **indices** of the two numbers such that they add up to a specific target.
 *
 * You may assume that each input would have **exactly** one solution, and you may not use the same element twice.
 *
 *
 * @author Ynjxsjmh
 * @email  ynjxsjmh@gmail.com
 * @since  2020-02-23 02:13
 * @tag    Array, Hash Table
 */

class Solution {
    /**
     * Approach 1: Brute Force
     * Time complexity: O(n^2)
     * Space complexity : O(1)
     */
    public int[] twoSum(int[] nums, int target) {
        for (int i=0; i<nums.length; i++) {
            for (int j=i+1; j<nums.length; j++) {
                if (nums[i]+nums[j]==target) {
                    return new int[]{i, j};
                }
            }
        }
        return new int[]{0, 0};
    }
}



// What is the best way to maintain a mapping of each element in the array to its index? A hash table.

class Solution {
    // Approach 2: Two-pass Hash Table
    // Time complexity : O(n)
    // Space complexity: O(n)
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }

        for (int i = 0; i < nums.length; i++) {
            int result = target - nums[i];
            if (map.containsKey(result) && map.get(result) != i) {
                // 后一个判断条件：Given nums = [3,2,4], target = 6
                return new int[] {i, map.get(result)};
            }
        }

        return new int[] {0, 0};
    }
}


